django/core/checks/settings.py

```diff
@@ -0,0 +1,119 @@
+    "SOUTH_DATABASE_ADAPTERS",
+    "SOUTH_AUTO_FREEZE_APP",
+    "SOUTH_TESTS_MIGRATE",
+    "SOUTH_LOGGING_ON",
+    "SOUTH_LOGGING_FILE",
+    "SOUTH_MIGRATION_MODULES",
+    "SOUTH_USE_PYC",
+    "TEST_CREATE",
+    "TEST_USER_CREATE",
+    "TEST_PASSWD",
+    "TEST_DATABASE_ENGINE",
+    "TEST_DATABASE_HOST",
+    "TEST_DATABASE_NAME",
+    "TEST_DATABASE_OPTIONS",
+    "TEST_DATABASE_PASSWORD",
+    "TEST_DATABASE_PORT",
+    "TEST_DATABASE_USER",
+    # Django 1.8
+    # https://docs.djangoproject.com/en/stable/releases/1.8/#features-removed-in-1-8
+    "SEND_BROKEN_LINK_EMAILS",
+    "CACHE_MIDDLEWARE_ANONYMOUS_ONLY",
+    # Django 1.9
+    # https://docs.djangoproject.com/en/stable/releases/1.9/#features-removed-in-1-9
+    # Django 1.10
+    # https://docs.djangoproject.com/en/stable/releases/1.10/#features-removed-in-1-10
+    "ALLOWED_INCLUDE_ROOTS",
+    "LOGOUT_URL",
+    "TEMPLATE_CONTEXT_PROCESSORS",
+    "TEMPLATE_DEBUG",
+    "TEMPLATE_DIRS",
+    "TEMPLATE_LOADERS",
+    "TEMPLATE_STRING_IF_INVALID",
+    # Django 2.0
+    # https://docs.djangoproject.com/en/stable/releases/2.0/#features-removed-in-2-0
+    "MIDDLEWARE_CLASSES",
+    # Django 2.1
+    # https://docs.djangoproject.com/en/stable/releases/2.1/#features-removed-in-2-1
+    "USE_ETAGS",
+    "SECURE_BROWSER_XSS_FILTER",
+    # Django 3.0
+    # https://docs.djangoproject.com/en/stable/releases/3.0/#features-removed-in-3-0
+    "DEFAULT_CONTENT_TYPE",
+    "PASSWORD_RESET_TIMEOUT_DAYS",
+    # Django 3.1
+    # https://docs.djangoproject.com/en/stable/releases/3.1/#features-removed-in-3-1
+    "DEFAULT_FILE_STORAGE",
+    "FILE_CHARSET",
+    # Django 4.0
+    # https://docs.djangoproject.com/en/stable/releases/4.0/#features-removed-in-4-0
+    "DEFAULT_HASHING_ALGORITHM",
+    "PASSWORD_RESET_TIMEOUT_DAYS",
+    "SECURE_BROWSER_XSS_FILTER",
+    # Django 4.1
+    # https://docs.djangoproject.com/en/stable/releases/4.1/#features-removed-in-4-1
+    # Django 5.0
+    # https://docs.djangoproject.com/en/stable/releases/5.0/#features-removed-in-5-0
+    "USE_L10N",
+    "USE_DEPRECATED_PYTZ",
+    "CSRF_COOKIE_MASKED",
+    "DATABASE_OPTIONS",
+    # Django 5.1
+    # https://docs.djangoproject.com/en/stable/releases/5.1/#features-removed-in-5-1
+    "DEFAULT_FILE_STORAGE",
+    "STATICFILES_STORAGE",
+}
+
+
+@register(Tags.settings)
+def check_deprecated_settings(**kwargs):
+    """
+    This check warns users who still use deprecated settings variables.
+    """
+
+    warnings = []
+    for attribute in dir(settings):
+        if attribute.isupper() and attribute in DEPRECATED_SETTINGS:
+            warnings.append(
+                Warning(
+                    f"You still use {attribute!r} in your Django settings file. "
+                    f"This attribute is deprecated.",
+                    hint="Please refer to the documentation and remove/replace "
+                    "this attribute.",
+                    id="settings.W001",
```

Comment:
```suggestion
                    id=f"settings.W001.{attribute}",
```
This might be a bit of strange idea but I'm wondering if we should add the setting to the id so that people can silence the check for a specific setting
This made me wonder if we should have silencing be a bit fancier so if I put `settings.W001` I want to silence `settings.W001.*`

Anyway, I was trying to think of nice ways folks who don't want to rename a setting can keep it (maybe @adamchainz might have some ideas)

Reply:
The ID should be a fixed string, not templated. The `obj` could be defined as the setting name.

[Ticket #26472](https://code.djangoproject.com/ticket/26472) tracks finer-grained silencing of system checks. When that gets added, it would allow silencing of this check.

But also, I don't think there's really going to be *any* use case for silencing this check, right? If you have such a setting, it's being ignored, and may as well be deleted, if not migrated.

Reply:
I think the use case is someone has created their own setting and it clashes (similar to [Tim's comment here](https://code.djangoproject.com/ticket/35674#comment:9))
You were using it for your own purposes, unaware it was ever used by Django core in the past

Reply:
It's true, that can happen. But is it so common that we need a solution for this? In the end, if somebody uses an already "taken" variable, renaming it is still an option - or he/she has to live with disabled checks. I guess it won't apply to more than 1% of all projects, maybe even less.

I'd argue with the paradigm "don't overengineer" - if people start complaining by the hundreds, we can always add this for 6.0, right?

Reply:
@sarahboyce Are we in agreement that this suggestion won't make it in?

Reply:
I agree that ticket-26472 would be the correct way to deal with this concern rather than my suggestion

Reply:
I had a thought about that topic. We could (!) group the dict by Django version and use this to make the message clearer.

```python
REMOVED_SETTINGS = {
    "4.0": {
        "MY_VARIABLE",
        "OTHER_VARIABLE",
    },
    "4.1": {
        ...
    }
}
```

...and then use this in the message:

```python
Warning(
    f"The {setting_name!r} setting was removed in {django_version} and its use is "
    f"not recommended.",
    hint="Please refer to the documentation and remove/replace "
    "this setting.",
    id="settings.W001",
                )
```

We'd also be able to add the changelog with a little more finesse... 🤔 

Any thoughts from the crowd?

Reply:
Yes, we should report the Django version that the setting was removed in.

Even better would be the version of deprecation, and then removal, like "Setting ... was deprecated in Django X and removed in Django Y". The release notes for the deprecated version explain the reason why, typically the removal version notes just say "... was removed.".

Reply:
If we have more than just the variable name, I would go with a dataclass containing the name, the deprecation version, the removal version and the changelog link. Is this something that fits to the existing coding patterns in Django core? If so, I'll prepare a little something. If not, I'd be happy about a different suggestion 🙂

Reply:
> Yes, we should report the Django version ...
> ... of deprecation, and then removal, like "Setting ... was deprecated in Django X and removed in Django Y".

These seems to me the most useful things to do.
Even better would be to add in the message also the URL to the deprecation and removal.

```
Setting ...
was deprecated in Django 1.5 (see https://docs.djangoproject.com/en/stable/releases/1.5/#features-deprecated-in-1-5)
and removed in Django 1.7 (see https://docs.djangoproject.com/en/stable/releases/1.7/#features-removed-in-1-7).
```